### PR TITLE
feat(mongo): division by zero and null in formulas TCTC-1552

### PR DESCRIPTION
**Pull request by davinov**
#992 reopened to resolve conflicts
Contrib by @CharlesRngrd (thanks a lot!)

### CODE REVIEW

File: src/lib/translators/mongo.ts [hunk 1]
```diff
@@ -1946,13 +1946,52 @@ export class Mongo36Translator extends BaseTranslator {
...
+          }
+          if (typeof obj[key] === 'object') {
+            iterate(obj[key]);
+          }
+        });
+      }
     };
+
+    // The first element was set for type reason
+    denominators.shift();
```
Comment (austil) [inline src/lib/translators/mongo.ts:1970]: #HowToTypeScript ? :sweat_smile: 
Can we get rid of this or are we really stuck with this code ?

File: src/lib/translators/mongo.ts [hunk 2]
```diff
@@ -1946,13 +1946,52 @@ export class Mongo36Translator extends BaseTranslator {
...
+
+    iterate(mongoFormulaTree);
```
Comment (austil) [inline src/lib/translators/mongo.ts:1972]: Could be renamed "collectDenominator" or something more explicit ?

File: src/lib/translators/mongo.ts [hunk 3]
```diff
@@ -1946,13 +1946,52 @@ export class Mongo36Translator extends BaseTranslator {
...
+        $addFields: {
+          [step.new_column]: mongoFormulaTree,
+        },
+      };
+    } else {
+      const conditions = [];
+
+      for (const d in denominators) {
+        conditions.push({ $in: [denominators[d], [0, null]] });
+      }
```
Comment (austil) [inline src/lib/translators/mongo.ts:1985]:
```suggestion
      const conditions = denominators.map((d) => ({ $in: [d, [0, null]] }));
```

Bit weird to use `for...in` for iterating over arrays, right ? :sweat_smile:

**Comment by davinov** [comment]
@austil I've taken your review as an opportunity to rework how we collect denominators so it doesn't introduce another iteration logic over formula trees twice.

2c72f2d 

and merged to fix changelog conflicts

**Review by austil — APPROVED** [review]
:green_circle:

**Pull request merged**